### PR TITLE
[DO NOT MERGE] Jest tests for composition

### PR DESCRIPTION
**Pull request by muenzpraeger**

### CODE REVIEW

File: force-app/main/default/lwc/compositionContactSearch/__tests__/compositionContactSearch.test.js [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+        inputFieldEl.dispatchEvent(new CustomEvent('change'));
+
+        jest.runAllTimers();
+
+        // Return a promise to wait for any asynchronous DOM updates. Jest
+        // will automatically wait for the Promise chain to complete before
+        // ending the test and fail the test if the promise ends in the
+        // rejected state
+        return Promise.resolve().then(() => {
+            // TODO rw: check for method call of findContacts
```
Comment (muenzpraeger) [inline force-app/main/default/lwc/compositionContactSearch/__tests__/compositionContactSearch.test.js]: @trevor-bliss @kv here I'm not getting return data from imperative Apex. I guess I'm doing the data mock further up wrong.

Comment (trevor-bliss) [reply force-app/main/default/lwc/compositionContactSearch/__tests__/compositionContactSearch.test.js]: I don't think the user "KV" is going to know the answer to this. @kevinv11n is probably interested though 😛

There's a lot of options and knobs to turn when it comes to mocking. Here's how I got it working:

// at top of test file - sets up initial mock to be shared between 
```
import findContacts from '@salesforce/apex/ContactController.findContacts';
jest.mock(
    '@salesforce/apex/ContactController.findContacts',
    () => {
        return {
            default: jest.fn()
        };
    },
    { virtual: true }
);
```

// in test case - sets up mock return value 
```
findContacts.mockResolvedValue([{ Id: '99', Name: 'Amy Taylor' }]);
```

// in test case - verify mock call
```
expect(findContacts).toHaveBeenCalledWith({ ...whatever expected shape is... });
```

File: force-app/main/default/lwc/compositionContactSearch/__tests__/compositionContactSearch.test.js [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        document.body.appendChild(element);
+
+        // Query lightning-input field element
+        const inputFieldEl = element.shadowRoot.querySelector(
+            'lightning-input'
+        );
+        inputFieldEl.value = 'Amy';
+        inputFieldEl.dispatchEvent(new CustomEvent('change'));
+
+        jest.runAllTimers();
```
Comment (trevor-bliss) [inline force-app/main/default/lwc/compositionContactSearch/__tests__/compositionContactSearch.test.js]: You need to call `jest.useFakeTimers()` somewhere before this, no?

File: force-app/main/default/lwc/compositionContactSearch/__tests__/compositionContactSearch.test.js [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+describe('c-composition-contact-search', () => {
+    afterEach(() => {
+        // The jsdom instance is shared across test cases in a single file so reset the DOM
+        while (document.body.firstChild) {
+            document.body.removeChild(document.body.firstChild);
+        }
+    });
+
+    it('renders one lightning-input field labelled "Search"', () => {
```
Comment (trevor-bliss) [inline force-app/main/default/lwc/compositionContactSearch/__tests__/compositionContactSearch.test.js]: labelled -> labeled

File: force-app/main/default/lwc/compositionContactSearch/__tests__/compositionContactSearch.test.js [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+            // TODO rw: wait for imperative Apex to be called with timer in mind
+            const contactTile = element.shadowRoot.querySelector(
+                'c-contact-tile'
+            );
+            expect(contactTile).not.toBeNull();
+            expect(contactTile.contact.Name).toBe('Amy Taylor');
+        });
+    });
+
+    it('renders the error panel', () => {
```
Comment (trevor-bliss) [inline force-app/main/default/lwc/compositionContactSearch/__tests__/compositionContactSearch.test.js]: Renders the error panel... when? Either add a `describe` block or expand this title to be more explicit for when the error panel should show.

File: force-app/main/default/lwc/compositionContactSearch/__tests__/compositionContactSearch.test.js (hunk 1 above)
Comment (muenzpraeger) [reply force-app/main/default/lwc/compositionContactSearch/__tests__/compositionContactSearch.test.js]: I applied this and it's throwing in the test run an error:

```
    TypeError: Cannot read property 'then' of undefined

      16 |         // eslint-disable-next-line @lwc/lwc/no-async-operation
      17 |         this.delayTimeout = setTimeout(() => {
    > 18 |             findContacts({ searchKey })
```

File: force-app/main/default/lwc/compositionContactSearch/__tests__/compositionContactSearch.test.js (hunk 2 above)
Comment (muenzpraeger) [reply force-app/main/default/lwc/compositionContactSearch/__tests__/compositionContactSearch.test.js]: Fixed

File: force-app/main/default/lwc/compositionContactSearch/__tests__/compositionContactSearch.test.js (hunk 3 above)
Comment (muenzpraeger) [reply force-app/main/default/lwc/compositionContactSearch/__tests__/compositionContactSearch.test.js]: Fixed

File: force-app/main/default/lwc/compositionContactSearch/__tests__/compositionContactSearch.test.js (hunk 4 above)
Comment (muenzpraeger) [reply force-app/main/default/lwc/compositionContactSearch/__tests__/compositionContactSearch.test.js]: Description updated

File: force-app/main/default/lwc/compositionContactSearch/__tests__/compositionContactSearch.test.js (hunk 1 above)
Comment (trevor-bliss) [reply force-app/main/default/lwc/compositionContactSearch/__tests__/compositionContactSearch.test.js]: You need to change `findContacts.mockResolvedValue = APEX_CONTACTS_SUCCESS;` to `findContacts.mockResolvedValue(APEX_CONTACTS_SUCCESS);`. 

The test will then fail with the message:
```
[{"Id": "99", "Name": "Amy Taylor"}]
    as argument 1, but it was called with
      {"searchKey": "Amy"}.
```

This is expected though if you look at how `findContacts` is called in `compositionContactSearch.js`. Based on the title "renders one contact tile based on user input", I don't think you should be verifying what `findContacts` is called with. That should be a separate test case.

Another thing to consider is that you'll need to wait an extra microtask for the DOM to rerender. The first microtask for the `findContacts` promise to resolve and set the `contacts` property, the second for the rerender.

Comment (muenzpraeger) [reply force-app/main/default/lwc/compositionContactSearch/__tests__/compositionContactSearch.test.js]: Updated. Although what's the recommendation for testing the `catch`? I tried `rejected`, but the c-error-panel doesn't get rendered.

https://github.com/trailheadapps/lwc-recipes/pull/33/files#diff-3a162ea4596850fe3ca4f34927482727R89

File: force-app/main/default/lwc/compositionContactSearch/__tests__/compositionContactSearch.test.js [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+        jest.runAllTimers();
+
+        // Return a promise to wait for any asynchronous DOM updates. Jest
+        // will automatically wait for the Promise chain to complete before
+        // ending the test and fail the test if the promise ends in the
+        // rejected state
+        return Promise.resolve()
+            .then(() => {
+                expect(findContacts()).rejects.toBe(APEX_CONTACTS_ERROR);
```
Comment (muenzpraeger) [inline force-app/main/default/lwc/compositionContactSearch/__tests__/compositionContactSearch.test.js]: @trevor-bliss Can you confirm that this is the correct way for mocking up the `catch` in the Apex promise? I'm asking b/c this returns true, but in the next microtask the errorPanelEl is always `undefined`.

Comment (muenzpraeger) [reply force-app/main/default/lwc/compositionContactSearch/__tests__/compositionContactSearch.test.js]: Great - fixed!

**Pull request merged**